src/nvim/os/fsnotification.c

```diff
@@ -0,0 +1,248 @@
+#include <stdbool.h>
+#include <stdlib.h>
+#include <assert.h>
+
+#include <uv.h>
+
+#include "nvim/vim.h"
+#include "nvim/buffer.h"
+#include "nvim/lib/khash.h"
+#include "nvim/lib/klist.h"
+#include "nvim/fileio.h"
+#include "nvim/os/fs_defs.h"
+#include "nvim/os/fsnotification.h"
+#include "nvim/os/event.h"
+
+
+// Apparentely, the destruction mechanism is not very well implemented in klist.
+// Proper resource handling will be done manually
+#define _destroy(x)
+KLIST_INIT(Watcher, Watcher, _destroy)
+#undef _destroy
+
+static klist_t(Watcher)* watchers_list = NULL;
+
+KHASH_MAP_INIT_STR(EventTable, int)
+khash_t(EventTable)* event_lookup = NULL;
+
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "os/fsnotification.c.generated.h"
+#endif
+
+typedef struct {
+    buf_T* buffer;
+    khiter_t lookup_iter;
+    uv_fs_event_t* handle;
+} EvtData;
+
+static void check_callback(Event evt)
+{
+  EvtData* data = (EvtData*) evt.data;
+  (void) buf_check_timestamp(data->buffer, false);
+
+  if ((kh_value(event_lookup, data->lookup_iter) & UV_RENAME) &&
+      os_file_exists(data->buffer->b_ffname)) {
+    fsnotification_delete_buffer(data->buffer);
+    fsnotification_add_buffer(data->buffer);
+  }
+
+  kh_value(event_lookup, data->lookup_iter) = 0;
+  free(data);
+}
+
+static void fs_event_callback(uv_fs_event_t* handle,
+                                    const char* filename,
+                                    int events,
+                                    int status)
+{
+  // we need to retrieve the full path for buflist_findname
+  char path[1024];
```

Comment:
This must be MAXPATHL unless you have good reasons to use 1024. If you do, it is good to document them here in a comment.